src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -80,19 +82,152 @@ def from_setting(d: dict) -> "SRAMParameters":
             mux=int(d["mux"])
         )
 
+
 Supply = NamedTuple('Supply', [
     ('name', str),
     ('pin', Optional[str]),
     ('tie', Optional[str]),
     ('weight', Optional[int])
 ])
 
-PinAssignment = NamedTuple('PinAssignment', [
+
+class PinAssignmentError(ValueError):
+    """Exception raised from parsing an invalid PinAssignment object."""
+
+
+class PinAssignmentSemiAutoError(PinAssignmentError):
+    """Exception raised from parsing a PinAssignment that requires semi-auto features without explicitly enabing
+    them."""
+
+
+class PinAssignmentPreplacedError(PinAssignmentError):
+    """Exception raised from parsing a replaced pin with extraneous information."""
+
+    def __init__(self, pin: "PinAssignment") -> None:
+        self.pin = pin
+
+    def __str__(self) -> str:
+        return "Pins {p} assigned as a preplaced pin with layers or side. Assuming pins are preplaced pins and ignoring layers and side.".format(
+            p=self.pin.pins)
+
+
+class PinAssignment(NamedTuple('PinAssignment', [
     ('pins', str),
     ('side', Optional[str]),
     ('layers', Optional[List[str]]),
-    ('preplaced', Optional[bool])
-])
+    ('preplaced', Optional[bool]),
```

Comment:
My understanding is that if `preplaced` no other options are valid (see the defaults.yml documentation)

Long term we may want to have subclasses of PinAssignment (e.g. `PreplacedPinAssignment`, etc)

Reply:
Preplaced means you only need the `pins` part of the struct because the pins are already defined by some internal structure so all other parameters are ignored.